colexec: add timestamp type

yuzefovich opened the pull request:
This PR is the rebased version of #40250 with a modification of how we're creating `DTimestamp` in `PhysicalTypeColElemToDatum`.

**colexec,coldata: support timestamp type**

This commit adds support for the TIMESTAMP type to the coldata and colexec
packages.

**colexec: add random projection tests for all types**

This test randomly runs binary comparisons against all types with random
data, verifying that the result of Datum.Compare matches the result of
the exec projection.

This found the bug with date infinity matching tracked in #40354, and
immediately found a bug with the timestamp implementation in the
previous commit, so I'm fairly sure it's a useful random test to have
around.

**colserde: add simple serialization for TIMESTAMP**

This commit adds simple (and quite inefficient) serialization
for TIMESTAMP type which uses the provided marshalling and
unmarshalling methods.

Release note: None

jordanlewis committed:
colexec,coldata: support timestamp type

This commit adds support for the TIMESTAMP type to the coldata and colexec
packages.

Release note: None

jordanlewis committed:
colexec: add random projection tests for all types

This test randomly runs binary comparisons against all types with random
data, verifying that the result of Datum.Compare matches the result of
the exec projection.

This found the bug with date infinity matching tracked in #40354, and
immediately found a bug with the timestamp implementation in the
previous commit, so I'm fairly sure it's a useful random test to have
around.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41662)
<!-- Reviewable:end -->

asubiotto commented:
Is this RFAL?

yuzefovich commented:
I want #41658 to go in first so that we can add the end-to-end tests for timestamp as well. Also, I'll need to address the comments from the old PR. So PTAL at #41658.

yuzefovich commented:
I addressed the remaining comments from the old PR.

asubiotto commented:
Perfect, will take a look after #41658 is in.

yuzefovich committed:
colserde: add simple serialization for TIMESTAMP

This commit adds simple (and quite inefficient) serialization
for TIMESTAMP type which uses the provided marshalling and
unmarshalling methods.

Release note: None

yuzefovich commented:
I've just added a simple serialization for TIMESTAMP. Here are the benchmarks:
```
BenchmarkArrowBatchConverter/Bool/nullFraction=0.00/BatchToArrow-8         	  500000	      2822 ns/op	 362.77 MB/s
BenchmarkArrowBatchConverter/Bool/nullFraction=0.25/BatchToArrow-8         	  500000	      2899 ns/op	 353.18 MB/s
BenchmarkArrowBatchConverter/Bool/nullFraction=0.50/BatchToArrow-8         	  500000	      2875 ns/op	 356.15 MB/s
BenchmarkArrowBatchConverter/Bool/nullFraction=0.00/ArrowToBatch-8         	 1000000	      2012 ns/op	 508.75 MB/s
BenchmarkArrowBatchConverter/Bool/nullFraction=0.25/ArrowToBatch-8         	 1000000	      1949 ns/op	 525.36 MB/s
BenchmarkArrowBatchConverter/Bool/nullFraction=0.50/ArrowToBatch-8         	 1000000	      1950 ns/op	 524.88 MB/s
BenchmarkArrowBatchConverter/Bytes/nullFraction=0.00/BatchToArrow-8        	   30000	     51289 ns/op	1277.76 MB/s
BenchmarkArrowBatchConverter/Bytes/nullFraction=0.25/BatchToArrow-8        	   30000	     51587 ns/op	1270.40 MB/s
BenchmarkArrowBatchConverter/Bytes/nullFraction=0.50/BatchToArrow-8        	   30000	     51618 ns/op	1269.61 MB/s
BenchmarkArrowBatchConverter/Bytes/nullFraction=0.00/ArrowToBatch-8        	  500000	      3117 ns/op	21024.69 MB/s
BenchmarkArrowBatchConverter/Bytes/nullFraction=0.25/ArrowToBatch-8        	  500000	      3049 ns/op	21490.23 MB/s
BenchmarkArrowBatchConverter/Bytes/nullFraction=0.50/ArrowToBatch-8        	  500000	      3054 ns/op	21456.54 MB/s
BenchmarkArrowBatchConverter/Int64/nullFraction=0.00/BatchToArrow-8        	10000000	       208 ns/op	39225.47 MB/s
BenchmarkArrowBatchConverter/Int64/nullFraction=0.25/BatchToArrow-8        	10000000	       217 ns/op	37582.33 MB/s
BenchmarkArrowBatchConverter/Int64/nullFraction=0.50/BatchToArrow-8        	10000000	       217 ns/op	37592.25 MB/s
BenchmarkArrowBatchConverter/Int64/nullFraction=0.00/ArrowToBatch-8        	10000000	       184 ns/op	44404.61 MB/s
BenchmarkArrowBatchConverter/Int64/nullFraction=0.25/ArrowToBatch-8        	 5000000	       248 ns/op	33021.16 MB/s
BenchmarkArrowBatchConverter/Int64/nullFraction=0.50/ArrowToBatch-8        	 5000000	       248 ns/op	32949.95 MB/s
BenchmarkArrowBatchConverter/Timestamp/nullFraction=0.00/BatchToArrow-8    	   20000	     74174 ns/op	 331.33 MB/s
BenchmarkArrowBatchConverter/Timestamp/nullFraction=0.25/BatchToArrow-8    	   20000	     74090 ns/op	 331.70 MB/s
BenchmarkArrowBatchConverter/Timestamp/nullFraction=0.50/BatchToArrow-8    	   20000	     73936 ns/op	 332.39 MB/s
BenchmarkArrowBatchConverter/Timestamp/nullFraction=0.00/ArrowToBatch-8    	  200000	     10489 ns/op	2342.86 MB/s
BenchmarkArrowBatchConverter/Timestamp/nullFraction=0.25/ArrowToBatch-8    	  200000	     10492 ns/op	2342.17 MB/s
BenchmarkArrowBatchConverter/Timestamp/nullFraction=0.50/ArrowToBatch-8    	  200000	     10523 ns/op	2335.36 MB/s
```
I marked each timestamp as 24 bytes since it has two int64 and a pointer. I think that it is better to have this simple serialization than not supporting it at all.

asubiotto reviewed (dismissed):
This all looks good as a first step but I'm slightly uncomfortable with our use of `time.Time`, mostly for serialization purposes. `arrow` implements a timestamp datatype (`arrow.TimestampType`: https://github.com/apache/arrow/blob/94a584bf483657d0bbf03b6090344a09b86fe5da/go/arrow/datatype_fixedwidth.go#L67) which I assume we would have to extend with comparison methods and I'm not sure how hard that would be, taking `Location` into account. That makes me realize that it doesn't seem like we've implemented the support for projections involving timestamps, e.g.:
```
root@127.0.0.1:54160/defaultdb> CREATE TABLE a (x TIMESTAMP);
CREATE TABLE

Time: 3.675ms

root@127.0.0.1:54160/defaultdb> SELECT x + interval '1 hour' FROM a;
```
Is that right? It will probably be worth it to add these.

> Reviewed 31 of 31 files at r1, 9 of 9 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-:-LsNanFJByM2CofmyVGn:b-p6176y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/col/colserde/arrowbatchconverter.go, line 264 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsN_i2T41pvLHtT8xiO:-LsN_i2T41pvLHtT8xiP:bnym79e) ([raw file](https://github.com/cockroachdb/cockroach/blob/15d7a13c3b73d2cb9b62ee352c609a59e63ea521/pkg/col/colserde/arrowbatchconverter.go#L264)):*
> ```Go
> 				bytesArr := array.NewBinaryData(d)
> 				bytes := bytesArr.ValueBytes()
> 				if bytes == nil {
> ```

Can we extract this and the same code in `coltypes.Bytes` into a helper function?
___
*[pkg/sql/colexec/mem_estimation.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNXDPN-0oPure-iLGS:-LsNXDPN-0oPure-iLGT:b9aalym) ([raw file](https://github.com/cockroachdb/cockroach/blob/523e77fdc20994453f6e3d6ea2ab5826c136970e/pkg/sql/colexec/mem_estimation.go#L59)):*
> ```Go
> 			acc += 50
> 		case coltypes.Timestamp:
> 			// time.Time has 2 int64s and a pointer.
> ```

Should we also account for a fully-set `Location` struct?
___
*[pkg/sql/colexec/projection_ops_test.go, line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNZrW02-mQW3HD3Mld:-LsNZrW02-mQW3HD3Mle:bvoot08) ([raw file](https://github.com/cockroachdb/cockroach/blob/b007f7178a330df2db8a949217c52c4bf8c3c560/pkg/sql/colexec/projection_ops_test.go#L188)):*
> ```Go
> 	const numTuples = 2048
> 	rng, _ := randutil.NewPseudoRand()
> 	evalCtx := tree.NewTestingEvalContext(cluster.MakeTestingClusterSettings())
> ```

nit: `defer evalCtx.Stop`?
___
*[pkg/sql/colexec/random_testutils.go, line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNajmC-zZOpCaDSymM:-LsNajmC-zZOpCaDSymN:bdsn0wy) ([raw file](https://github.com/cockroachdb/cockroach/blob/15d7a13c3b73d2cb9b62ee352c609a59e63ea521/pkg/sql/colexec/random_testutils.go#L104)):*
> ```Go
> 		timestamps := vec.Timestamp()
> 		for i := 0; i < n; i++ {
> 			timestamps[i] = timeutil.Unix(rng.Int63n(1000000), rng.Int63n(1000000))
> ```

I wonder if we can find a way to also randomize `Location`?
___
*[pkg/sql/colexec/selection_ops_tmpl.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNXVb22j272MMjT3UW:-LsNXVb22j272MMjT3UX:bsbydg) ([raw file](https://github.com/cockroachdb/cockroach/blob/523e77fdc20994453f6e3d6ea2ab5826c136970e/pkg/sql/colexec/selection_ops_tmpl.go#L57)):*
> ```Go
> 
> // Dummy import to pull in "time" package.
> var _ = time.Day
> ```

nit: for consistency's sake, this and other uses of `time.Day` should be `time.Time` or vice-versa


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
You're right that we don't implement any projections because timestamps have binary operators only with other types (like intervals) which we don't support.

I feel like using the existing marshaling methods for the serialization is less error-prone, and if we later decide to switch representation, we (similarly to UUID discussion) will bump the DistSQL server version, so I'm leaning towards going forward with the current approach since it seems to work with a reasonable efficiency without needing to spend extra time figuring out the arrow timestamp type.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-:-LsOdpPI2kJtoD2jKLBj:b-dzznn0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/col/colserde/arrowbatchconverter.go, line 264 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsN_i2T41pvLHtT8xiO:-LsODw1d9RaVE44Yriad:bna8ydl) ([raw file](https://github.com/cockroachdb/cockroach/blob/15d7a13c3b73d2cb9b62ee352c609a59e63ea521/pkg/col/colserde/arrowbatchconverter.go#L264)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Can we extract this and the same code in `coltypes.Bytes` into a helper function?
</blockquote></details>

Now that the flat bytes implementation is in, there is no duplication.
___
*[pkg/sql/colexec/mem_estimation.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNXDPN-0oPure-iLGS:-LsOdSGP-HYQh3Q3enEz:b9o7jmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/523e77fdc20994453f6e3d6ea2ab5826c136970e/pkg/sql/colexec/mem_estimation.go#L59)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Should we also account for a fully-set `Location` struct?
</blockquote></details>

Hm, I'm not sure. I feel like these `Location`s should be cached, but I haven't found the code to support this feeling. If my feeling is correct, then accounting for the whole `Location` struct on every timestamp will be overaccounting.
___
*[pkg/sql/colexec/projection_ops_test.go, line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNZrW02-mQW3HD3Mld:-LsOFBDgBqjzU0Td44l4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b007f7178a330df2db8a949217c52c4bf8c3c560/pkg/sql/colexec/projection_ops_test.go#L188)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: `defer evalCtx.Stop`?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/random_testutils.go, line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNajmC-zZOpCaDSymM:-LsOd3g393cM3hrT_wgS:bxxh6pb) ([raw file](https://github.com/cockroachdb/cockroach/blob/15d7a13c3b73d2cb9b62ee352c609a59e63ea521/pkg/sql/colexec/random_testutils.go#L104)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I wonder if we can find a way to also randomize `Location`?
</blockquote></details>

Done. This increases the time to run this test to about 2 seconds, but I guess it's ok.
___
*[pkg/sql/colexec/selection_ops_tmpl.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNXVb22j272MMjT3UW:-LsOdHwe7ROaGeUO02ED:b-xiw2bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/523e77fdc20994453f6e3d6ea2ab5826c136970e/pkg/sql/colexec/selection_ops_tmpl.go#L57)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: for consistency's sake, this and other uses of `time.Day` should be `time.Time` or vice-versa
</blockquote></details>

Nice catch, done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
OK. Could you create issues to track the addition of interval types and projections with timestamps?

Sounds good, let's move forward with this approach.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> minus `Location` generation

> Reviewed 14 of 14 files at r4, 9 of 9 files at r5, 10 of 10 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-:-LsSNoC28GE3rJBh7Yud:b-is9mtv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/colexec/mem_estimation.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNXDPN-0oPure-iLGS:-LsSO1eD3plbpY3ZSJbl:bsquwje) ([raw file](https://github.com/cockroachdb/cockroach/blob/523e77fdc20994453f6e3d6ea2ab5826c136970e/pkg/sql/colexec/mem_estimation.go#L59)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, I'm not sure. I feel like these `Location`s should be cached, but I haven't found the code to support this feeling. If my feeling is correct, then accounting for the whole `Location` struct on every timestamp will be overaccounting.
</blockquote></details>

Oh, I see. Could you add a comment?
___
*[pkg/sql/colexec/random_testutils.go, line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsSPjIQ9cbphhVS-mve:-LsSPjIQ9cbphhVS-mvf:b3ucjrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/69245a31921f5b1ef8dd2e359e9370949c758385/pkg/sql/colexec/random_testutils.go#L47)):*
> ```Go
> 		for b := byte('A'); b <= 'Z'; b++ {
> 			for c := byte('A'); c <= 'Z'; c++ {
> 				loc, err := timeutil.LoadLocation(string([]byte{a, b, c}))
> ```

Instead of doing this, I would choose a couple of random locations from this list:  https://en.wikipedia.org/wiki/List_of_tz_database_time_zones


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-:-LsT_7eJ8iwY0S2ntY5F:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/mem_estimation.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsNXDPN-0oPure-iLGS:-LsTIdzaAT3fcZZn_w1q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/523e77fdc20994453f6e3d6ea2ab5826c136970e/pkg/sql/colexec/mem_estimation.go#L59)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Oh, I see. Could you add a comment?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/random_testutils.go, line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/41662#-LsSPjIQ9cbphhVS-mve:-LsTJpJxDtjfokayYqr6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69245a31921f5b1ef8dd2e359e9370949c758385/pkg/sql/colexec/random_testutils.go#L47)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Instead of doing this, I would choose a couple of random locations from this list:  https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Seems like unrelated lint flake.

bors r+

asubiotto commented:
I think we've now got a green light to backport this (as well as the UUID and IS NOT NULL PRs)
